File: packages/polaris/src/modules/client-registry.ts
import Ajv from "ajv";
import { completionsPost, getToolCall } from "@/modules/ai/completions";

type Json = null | boolean | number | string | Json[] | { [k: string]: Json };

type ExecContext = {
    inputs: Record<string, Json>;
    state: Record<string, Json>;
    nodeId: string;
    graphId: string;
    graph: any;
};

type ApiCallSpec = {
    target: string;
    input?: Json;
};

type OpResult = { ok: true; result: Json } | { ok: false; error: { code: string; message: string } };

export class ClientRegistry {
    private aiConfig: { aiBaseUrl: string; aiApiKey: string; aiModel: string };
    private ajv: Ajv;
    private apiTargets: Record<string, (input: any) => Promise<any>>;
    readonly capabilities: string[];
    constructor(aiConfig: { aiBaseUrl: string; aiApiKey: string; aiModel: string }) {
        this.aiConfig = aiConfig;
        this.ajv = new Ajv();
        this.capabilities = ["llm", "galaxy.read"];
        this.apiTargets = {
            "galaxy.history.list": async (input) => {
                const params = new URLSearchParams();
                if (input?.limit) {
                    params.set("limit", String(input.limit));
                }
                const res = await fetch(`/api/histories?${params.toString()}`, { credentials: "same-origin" });
                if (!res.ok) {
                    throw new Error(`HTTP ${res.status}`);
                }
                return await res.json();
            },
            "galaxy.history.contents": async (input) => {
                const res = await fetch(`/api/histories/${input.history_id}/contents`, { credentials: "same-origin" });
                if (!res.ok) {
                    throw new Error(`HTTP ${res.status}`);
                }
                return await res.json();
            },
            "galaxy.dataset.show": async (input) => {
                const id = input?.dataset_id;
                if (!id) {
                    throw new Error("dataset_id missing");
                }
                const res = await fetch(`/api/datasets/${id}`, { credentials: "same-origin" });
                if (!res.ok) {
                    throw new Error(`HTTP ${res.status}`);
                }
                return await res.json();
            },
        };
    }
    async plan(ctx: ExecContext, spec: { prompt?: string; node: any; tools: any[]; outputSchema: any }) {
        const systemPrompt =
            spec.prompt || "You are a routing component. You MUST call the provided tool. Do not respond with text.";
        const messages = [{ role: "system", content: systemPrompt }, ...this.sanitize(ctx.inputs.transcripts as any[])];
        const tools = this.buildRouteTool(ctx, spec.node, spec.outputSchema);
        const toolName = tools[0].function.name;
        const reply = await completionsPost({
            ...this.aiConfig,
            messages,
            tools,
            tool_choice: {
                type: "function",
                function: { name: toolName },
            },
        });
        const call = getToolCall(toolName, reply?.choices?.[0]?.message?.tool_calls);
        if (!call) {
            throw new Error("planner did not produce tool call");
        }
        if (spec.outputSchema) {
            const validate = this.ajv.compile(spec.outputSchema);
            if (!validate(call)) {
                throw new Error(`planner output schema violation: ${this.ajv.errorsText(validate.errors)}`);
            }
        }
        return call;
    }

    async callApi(ctx: ExecContext, spec: ApiCallSpec): Promise<OpResult> {
        const fn = this.apiTargets[spec.target];
        if (!fn) {
            return { ok: false, error: { code: "unknown_api_target", message: spec.target } };
        }
        try {
            const result = await fn(spec.input ?? {});
            return { ok: true, result };
        } catch (e: any) {
            return {
                ok: false,
                error: { code: "api_call_failed", message: String(e?.message ?? e) },
            };
        }
    }
    private sanitize(transcripts: any[]) {
        if (!Array.isArray(transcripts)) {
            return [];
        }
        return transcripts
            .filter((t) => typeof t?.content === "string" && t.content.length > 0)
            .map((t) => ({ role: t.role, content: t.content }));
    }

    private buildRouteTool(ctx: ExecContext, node: any, outputSchema?: any) {
        let nextEnum: string[];
        if (
            outputSchema &&
            outputSchema.properties &&
            outputSchema.properties.next &&
            Array.isArray(outputSchema.properties.next.enum)
        ) {
            nextEnum = outputSchema.properties.next.enum;
        } else {
            nextEnum = Object.keys(ctx.graph.nodes);
        }
        const properties: any = {
            next: {
                type: "string",
                enum: nextEnum,
            },
        };
        const required = new Set<string>(["next"]);
        if (node.enum_from && outputSchema && Array.isArray(outputSchema.required)) {
            const src = ctx.state[node.enum_from.state];
            if (!Array.isArray(src)) {
                throw new Error(`enum_from source is not an array: ${node.enum_from.state}`);
            }
            let values = src;
            if (node.enum_from.filter) {
                values = values.filter((v: any) => v?.[node.enum_from.filter.field] === node.enum_from.filter.equals);
            }
            const enumValues = values
                .map((v: any) => v?.[node.enum_from.field])
                .filter((v: any) => typeof v === "string");
            const field = outputSchema.required.find((k: string) => k !== "next");
            if (field) {
                properties[field] = {
                    type: "string",
                    enum: enumValues,
                };
                required.add(field);
            }
        }
        return [
            {
                type: "function",
                function: {
                    name: "route",
                    description: "Select the next node and required identifiers.",
                    parameters: {
                        type: "object",
                        required: Array.from(required),
                        properties,
                        additionalProperties: false,
                    },
                },
            },
        ];
    }
}

File: packages/polaris/polaris/polaris/client_registry.py
from jsonschema import Draft7Validator
import json


class ClientRegistry:
    def __init__(self, ai_config):
        self.ai_config = ai_config
        self.capabilities = ["llm", "galaxy.read"]
        self.api_targets = {
            "galaxy.history.list": self._history_list,
            "galaxy.history.contents": self._history_contents,
            "galaxy.dataset.show": self._dataset_show,
        }

    # ---------- Planner ----------

    async def plan(self, ctx, spec):
        system_prompt = spec.get(
            "prompt",
            "You are a routing component. You MUST call the provided tool. Do not respond with text."
        )

        messages = [{"role": "system", "content": system_prompt}]
        messages.extend(self.sanitize(ctx["inputs"].get("transcripts")))

        tools = self.build_route_tool(ctx, spec["node"], spec.get("outputSchema"))
        tool_name = tools[0]["function"]["name"]

        reply = await completions_post({
            **self.ai_config,
            "messages": messages,
            "tools": tools,
            "tool_choice": {
                "type": "function",
                "function": {"name": tool_name},
            },
        })

        call = get_tool_call(
            tool_name,
            reply.get("choices", [{}])[0]
                .get("message", {})
                .get("tool_calls")
        )

        if not call:
            raise Exception("planner did not produce tool call")

        arguments = call["function"]["arguments"]
        if isinstance(arguments, str):
            arguments = json.loads(arguments)

        if spec.get("outputSchema"):
            validator = Draft7Validator(spec["outputSchema"])
            errors = list(validator.iter_errors(arguments))
            if errors:
                raise Exception(
                    "planner output schema violation: "
                    + "; ".join(e.message for e in errors)
                )

        return arguments

    # ---------- API Calls ----------

    async def call_api(self, ctx, spec):
        fn = self.api_targets.get(spec["target"])
        if not fn:
            return {
                "ok": False,
                "error": {
                    "code": "unknown_api_target",
                    "message": spec["target"],
                },
            }
        try:
            result = await fn(spec.get("input", {}))
            return {"ok": True, "result": result}
        except Exception as e:
            return {
                "ok": False,
                "error": {
                    "code": "api_call_failed",
                    "message": str(e),
                },
            }

    # ---------- Sanitization ----------

    def sanitize(self, transcripts):
        if not isinstance(transcripts, list):
            return []
        out = []
        for t in transcripts:
            content = t.get("content")
            if isinstance(content, str) and content:
                out.append({"role": t.get("role"), "content": content})
        return out

    # ---------- Tool Builder ----------

    def build_route_tool(self, ctx, node, output_schema=None):
        if (
            output_schema
            and output_schema.get("properties", {}).get("next", {}).get("enum")
        ):
            next_enum = output_schema["properties"]["next"]["enum"]
        else:
            next_enum = list(ctx["graph"]["nodes"].keys())

        properties = {
            "next": {
                "type": "string",
                "enum": next_enum,
            }
        }

        required = {"next"}

        enum_from = node.get("enum_from")
        if enum_from and output_schema and isinstance(output_schema.get("required"), list):
            src = ctx["state"].get(enum_from["state"])
            if not isinstance(src, list):
                raise Exception(
                    f"enum_from source is not an array: {enum_from['state']}"
                )

            values = src
            filt = enum_from.get("filter")
            if filt:
                values = [
                    v for v in values
                    if v.get(filt["field"]) == filt["equals"]
                ]

            enum_values = [
                v.get(enum_from["field"])
                for v in values
                if isinstance(v.get(enum_from["field"]), str)
            ]

            field = next(k for k in output_schema["required"] if k != "next")

            properties[field] = {
                "type": "string",
                "enum": enum_values,
            }
            required.add(field)

        return [
            {
                "type": "function",
                "function": {
                    "name": "route",
                    "description": "Select the next node and required identifiers.",
                    "parameters": {
                        "type": "object",
                        "required": list(required),
                        "properties": properties,
                        "additionalProperties": False,
                    },
                },
            }
        ]

    # ---------- Galaxy API stubs ----------

    async def _history_list(self, input):
        limit = input.get("limit")
        params = f"?limit={limit}" if limit else ""
        res = await fetch(f"/api/histories{params}")
        return await res.json()

    async def _history_contents(self, input):
        history_id = input["history_id"]
        res = await fetch(f"/api/histories/{history_id}/contents")
        return await res.json()

    async def _dataset_show(self, input):
        dataset_id = input.get("dataset_id")
        if not dataset_id:
            raise Exception("dataset_id missing")
        res = await fetch(f"/api/datasets/{dataset_id}")
        return await res.json()
